docs/langref.md

```diff
@@ -442,6 +442,32 @@ Example := Rectangle {
 }
 ```
 
+## Purity
+
+Slint's property evaluation is lazy and "reactive", which means that property bindings are only evaluated when they are used, and the value is cached.
+If any of the dependent properties change, the binding will be re-evaluated the next time the property is queried.
+Binding evaluation should be "pure", meaning that it should not have any side effects.
+For example, evaluating a binding should not change any other properties.
+Side effects are problematic because it is not always clear when they will happen.
+Lazy evaluation may change their order or affect whether they happen at all.
+In addition, changes to properties while their binding is being evaluated may result in unexpected behavior.
+
+For this reason, binding are required to be pure. And we check that no side effect operation are done in pure context.
+Pure context includes binding expression, body of pure functions, body of pure callback handler.
+Un such contect, it is not allowed to change a property, or call a non-pure callback or function.
+
+Callbacks and public function can be annotated with the `pure` keyword.
+Private function can also be annotated with `pure`, but its purity would aotimatically be infered if it is not
+annotated.
```

Comment:
```suggestion
For this reason, bindings are required to be pure. The Slint compiler checks that code in pure contexts has no side effects.
Pure contexts includes binding expressions, bodies of pure functions, and bodies of pure callback handlers.
In such a context, it is not allowed to change a property, or call a non-pure callback or function.

Callbacks and public functions can be annotated with the `pure` keyword.
Private functions can also be annotated with `pure`, otherwise their purity is automatically inferred.
```